Add support for more expressive span status API (#123)
open-telemetry/oteps: text/0123-improve-span-status-api.md

# Improve Span Status API

Allow the Span Status API to represent more kinds of status

## Motivation

Right now OpenTelemetry Status is defined as an enumeration of gRPC status codes. Although I couldn't find design criteria written down for this API I fear it is too narrowly defined to be useful across the full breadth of scenarios OpenTelemetry targets.

> OpenTelemetry allows Spans to be created to represent any operation, including those which don't involve communication with another component (Kind = Internal). These underlying operations can have native status representations from a particular domain or a language such as POSIX status codes, HRESULTs, many variations of exceptions, error messages, HTTP status, or gRPC status. However to capture the status as part of an OpenTelemetry span it must first be mapped to something in OpenTelemetry's object model and this mapping has the potential to create a few problems:

Reviewer: Please add line breaks at least every sentence, otherwise it is impossible to comment on specific parts of the paragraph.

Author: Yep, that was my bad, getting to used to writing my markdown with typora, will do : )

## Explanation

#### Data representation of Span Status

We should aim for representations that require minimal-to-no end-developer translation effort from the native representation and capture at least the key numeric/string data developers are likely to search for or relate to. I suggest status is this set of information:

1. StatusType - The name of the type of error such as "HTTP", "gRPC", "LanguageException", "HRESULT", "POSIX", or "ErrorMessage". The list is end-user-extensible but common status type names should be standardized. (Perhaps there is already some standardization we could borrow?)
2. StatusData - A discriminated union of:
   - An integer, a string, or a tuple of integer and string. These options can be used for:
     - Enumerated status codes: For example an http status code could be represented as 404, "Not Found", or both. In the case of common status codes OpenTelemetry SDK or backend could optionally assist in filling out the remainder of a partially specified enumeration value. For enumerations that aren't well-known the community of users is responsible for determining any conventions.

>      - Free-form error messages

Reviewer 2: How would a named error with a separate message be captured? I would treat these separately and not combine them in one free-form string. This would allow for searching, filtering and grouping by the error name alone but still preserve the message. A backend could, for example, display the number of all `EntityNotFound` errors in one place or filter for them in another place where users can inspect the messages bearing the entities that were not found.

>      - Free-form error messages

Reviewer 2: How would a named error with a separate message be captured? I would treat these separately and not combine them in one free-form string. This would allow for searching, filtering and grouping by the error name alone but still preserve the message. A backend could, for example, display the number of all `EntityNotFound` errors in one place and filter for them in another place where users can inspect the messages bearing the entities that were not found.

Author: The current proposal wouldn't be able to capture this, but we could change the proposal to allow it. I've added some text to the trade-offs and mitigations section mentioning the idea. I'd be fine adding it, not sure if we'll see additional opinions about it?

Reviewer 2: Personally, I'd add it right from the start. I'd expect users want to capture the information in a way that most closely resembles how it's reported by their respective language or framework, rather than having to combine both in one string that potentially has to be parsed/decomposed for analysis later again.

Author: My original intent wasn't that users would combine strings, but that they would provide either an enumerated status code OR a message, not both. If we anticipate it is common to provide both I agree I'd rather provide an API that easily facilitates that vs. an API that encourages users to create private conventions to merge and split text.

Reviewer 2: I expect it to be common enough to support it in the API. Often it will indeed be a type of error that can be enumerated but will not necessarily have an integer representation like we have in HTTP but rather a name of the error (like _"Entity not found"_ or an exception name) with a (detailed) description (e.g. _"The entity with ID '7583' was not found in the table 'products'"_ or an exception message).
Therefore I'm suggesting to have two strings called, for example, _name_, _brief_ or _kind_ and a separate string called _details_, _message_ or _description_.
Happy to hear others' opinions on that.

   - Exception object - whatever the SDK language's default exception datatype is, if it has one.
   - void
3. SuccessHint - An optional boolean that represents the span author's best guess whether this status represents a successful or failed operation, however they choose to define those terms. For well-known status types I'd suggest the hint be ignored but for user-defined status types this is likely the only clue whether the span should be surfaced in a UI as being abnormal or failed.

> Although UI creators are free experiment with how the data is presented I expect most presentations would either be the StatusData alone, or the StatusData qualified with the StatusType and some separator character. For example StringData alone might create names like "FileNotFoundException", "503", "E_FAIL (0x80004005)", "SyntaxError on line 405: Did you forget a semicolon?", and "BadQuery".

Reviewer 2: ```suggestion
Although UI creators are free experiment with how the data is presented I expect most presentations would either be the StatusData alone, or the StatusData qualified with the StatusType and some separator character. For example StatusData alone might create names like "FileNotFoundException", "503", "E_FAIL (0x80004005)", "SyntaxError on line 405: Did you forget a semicolon?", and "BadQuery".
```

> Although UI creators are free experiment with how the data is presented I expect most presentations would either be the StatusData alone, or the StatusData qualified with the StatusType and some separator character. For example StringData alone might create names like "FileNotFoundException", "503", "E_FAIL (0x80004005)", "SyntaxError on line 405: Did you forget a semicolon?", and "BadQuery".

Reviewer 2: ```suggestion
Although UI creators are free experiment with how the data is presented I expect most presentations would either be the StatusData alone, or the StatusData qualified with the StatusType and some separator character. For example StatusData alone might create names like "FileNotFoundException", "503", "E_FAIL (0x80004005)", "SyntaxError on line 405: Did you forget a semicolon?", and "BadQuery".
```

Author: Thanks!

Exceptions could have progressive level of detail drilling into messages, stack traces, inner exceptions, links to source, etc if the exporter serialized sufficient data but how and whether that occurs is out-of-scope in this design. 

#### API to capture this data

I would suggest an API called Span.SetStatus(...) that takes all the arguments above, and optionally overloads or default parameters that make common calls easier. For example in C#:

````C#

> void SetStatus(Exception spanException, string statusType = "LanguageException", bool? successHint = null);

Reviewer: Capturing an exception with traceback and storing it until export is potentially extremely memory heavy (e.g. in Python the exception has a reference to a traceback, and the traceback has references to all local variables on the stack, preventing them from being GCd).

So the solution to this would probably to call into some exporter-specific code already when calling this API.

Author: Thanks! I added a comment about this issue and potential solution to the internal details section.

- I expect developers both at the time they are emitting trace data and when viewing that trace data would find it is idiomatic to  include the status of the Span's workload together with the description of the Span. Identifying failed or abnormal spans is a typical APM operation.

**Closed vs. open-ended status descriptions** - We could make status represented by a fixed set of options such as the current gRPC codes or any larger/different set. It likely is simpler for tools consuming the data to only operate on fixed options and it is pleasant to imagine that all (or most) status results will be reasonably categorized into the fixed list. However I do not expect most users to take up the responsibility of mapping their status results if OpenTelemetry/back-end tools don't provide a clear representation. Instead these users will not set status or choose options that are generic and inconsistent rendering the data useless to everyone. For a fixed list to work I believe we either have to curate a very large list of options or accept that significant portions of the developer audience will not find the status representation useful.

> **API using attribute semantic conventions ** - It is also possible to do this via semantic conventions on attributes and although I think the strongly typed API is preferable I don't have that strong of an opinion. Semantic conventions are likely to have higher performance overhead, higher risk of error in key names and are less discoverable/refactorable in IDEs. The advantages are that there is some past precedent for doing this specific to http codes and new conventions can be added easily. If we go semantic conventions it does imply that Exception becomes a type that can be directly passed as an argument to SetAttribute(). Requiring the user to destructure the exception into a list of key value pairs would be overly onerous and error-prone for a common usage scenario. If desired the SDK or exporter could destructure it, but that can be determined independently from API design and I'd like to keep it out of scope.

Reviewer: Note that you can provide a nicely typed API and still use semantic conventions. E.g. the API could just "serialize" whatever you give it into SetAttribute / CreateEvent calls. The difference is that the API between the exporter and the API is then "weakly typed".

To point at another project, the idea of "typed spans", i.e. utilities to have strongly typed APIs for semantic conventions, is at least as old as OpenTelemetry. E.g. see "Typed Spans" in the PR description of https://github.com/open-telemetry/opentelemetry-specification/pull/571

Reviewer: As an additional advantage, the exception APIs could be provided as a package separate from the API, (you could provide a function `com.opentelemetry.contrib.errorutils.ErrorUtils.captureError(Span, other params)` that sets any attributes/events on the span).

Author: Thanks! I've added this suggestion to the list of alternatives.

> **API using attribute semantic conventions ** - It is also possible to do this via semantic conventions on attributes and although I think the strongly typed API is preferable I don't have that strong of an opinion. Semantic conventions are likely to have higher performance overhead, higher risk of error in key names and are less discoverable/refactorable in IDEs. The advantages are that there is some past precedent for doing this specific to http codes and new conventions can be added easily. If we go semantic conventions it does imply that Exception becomes a type that can be directly passed as an argument to SetAttribute(). Requiring the user to destructure the exception into a list of key value pairs would be overly onerous and error-prone for a common usage scenario. If desired the SDK or exporter could destructure it, but that can be determined independently from API design and I'd like to keep it out of scope.

Author: (I love reading paragraphs hunting for the changed space character! ... agreed and fixed 👍 )

> **API using Span events** - Most of the rationale for attribute semantic conventions also applies here, events are effectively another key-value store mechanism. The timestamp that is attached to an event appears to hold little value as status is probably produced at the same approximate time the span end timestamp is recorded. Similar to attribute conventions it sounds like there is precedent for storing some errors as events.

Author: I think we are in agreement? Perhaps my second paragraph made it sound like events aren't semantic conventions so I've clarified in the title that events are also conventions.

I've been limiting the scope of the discussion to the tracing API so here I was attempting to distinguish whether the user provides the information via strongly typed API or provides it via loosely typed conventions. I agree that a strongly typed API could store the data internally as attributes/events and doing so probably (but not necessarily) has some implications on the SDK API for reading the data back.

> **API using Span events** - Most of the rationale for attribute semantic conventions also applies here, events are effectively another key-value store mechanism. The timestamp that is attached to an event appears to hold little value as status is probably produced at the same approximate time the span end timestamp is recorded. Similar to attribute conventions it sounds like there is precedent for storing some errors as events.

Author: I added some text in the implementation section of the coming revision to explicitly call out the ability to store the data into attributes or events regardless of the tracing API that provides the data to us.

codes. Although I couldn't find design criteria written down for this API
I fear it is too narrowly defined to be useful across the full breadth of
scenarios OpenTelemetry targets.

OpenTelemetry allows Spans to be created to represent any operation, including
those which don't involve communication with another component (Kind =
Internal). These underlying operations can have native status representations
from a particular domain or a language such as POSIX status codes, HRESULTs,
many variations of exceptions, error messages, HTTP status, or gRPC status.
However to capture the status as part of an OpenTelemetry span it must first be
mapped to something in OpenTelemetry's object model and this mapping has the
potential to create a few problems:

- **Inconsistency** - If the mapping from native representation to
OpenTelemetry representation isn't well-defined then API users or SDK
implementations are unlikely to choose the same mapping. This makes collected
data hard to work with because it can no longer be treated uniformly.
- **Loss of fidelity** - Mapping from a status representation that
distinguishes a large number of different results to one that only

> distinguishes a few is inherently lossy. This prevents users from isolating

Reviewer 3: My understanding of the span status code is slightly different than the summary in this document. I see the span status code as a categorization of the status meant to augment our understanding of the span's condition. I would not characterize the span status code as a lossy transformation: I'd describe it as additive.

As an example, suppose we have an HTTP server and it encounters some kind of condition. In the HTTP domain we have well-defined status codes, which can be encoded in a non-lossy way using the [`http.status_code` semantic convention](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md#status). 

I believe one of the sources of confusion here is that in order to categorize an error into one of the canonical span status codes, the author has to take a position, and that position will not be unique. Depending on positionality, different users will categorize the various HTTP status codes differently. Accordingly, I would define the span status code as the span actor's perspective on the activity it performed, whether how it would categorize its own status. Other actors in a system may witness consequential failures (e.g., a client receiving a HTTP 500 code) and handle them in a way that disagrees with the self-assessment used by a particular span. E.g., a server may decide that a 500 status is maps to **Internal** errors, and the client that receives a 500 may decide to retry, thus suppressing the 500. That 500 could become a span event with `http.status_code` and `http.status_text` attributes, and still the client could have a successful retry and set the span status to **OK**.

Summarizing, I don't see a problem with Status code as it stands, but think we need to recognize that the Status code is not a unique position.

Author: Thanks for the thoughtful feedback!

> Summarizing, I don't see a problem with Status code as it stands

 I'm not sure how to relate your opinion on the status code concept to the OTEP overall. How do you feel about the goals outlined in the goals section of the OTEP:
a) They aren't useful and don't need to be pursued?
b) They are useful but the existing Span API already satisfies them?
c) They are useful, and the existing Span API doesn't satisfy them. However we shouldn't frame the proposal as a change to the 'Status' API because that API is already fulfilling a different important purpose
d) Something else?

> I believe one of the sources of confusion here is that in order to categorize an error into one of the canonical span status codes, the author has to take a position, and that position will not be unique

That sounds like the problem I named **Inconsistency** in the text, yes? "If the mapping from native representation to
OpenTelemetry representation isn't well-defined then API users or SDK implementations are unlikely to choose the same mapping."

guess whether this status represents a successful or failed operation, however
they choose to define those terms. For well-known status types I'd suggest the
hint be ignored but for user-defined status types this is likely the only clue
whether the span should be surfaced in a UI as being abnormal or failed.

Although UI creators are free experiment with how the data is presented I
expect most presentations would either be the StatusData alone, or the
StatusData qualified with the StatusType and some separator character. For
example StatusData alone might create names like "FileNotFoundException",
"503", "E_FAIL (0x80004005)", "SyntaxError on line 405: Did you forget a
semicolon?", and "BadQuery".

Exceptions could have progressive level of detail drilling into messages,
stack traces, inner exceptions, links to source, etc if the exporter serialized
sufficient data but how and whether that occurs is out-of-scope in this design. 

#### API to capture this data

I would suggest an API called Span.SetStatus(...) that takes all the arguments
above, and optionally overloads or default parameters that make common calls
easier. For example in C#:

````C#

> void SetStatus(Exception spanException, string statusType = "LanguageException", bool? successHint = null);

Reviewer 4: Multiple exceptions can occur during the lifetime of a span and experiencing an exception may or may not result in a non-ok status depending on the circumstances. How can multiple exceptions can handled using this API?

Reviewer: Multiple functions returning error codes may be called during the lifetime of a span and an error return may result in a non-ok status or not.

I think it's up to the user/instrumentation to decide which exception was the semantically important one. Most likely the one that went uncaugth.

Author: Agreed with @Oberon00. I also attempted to capture this concept in the scoping section above but perhaps the wording wasn't good?

> The operation represented by a span might be composed of sub-operations (for
> example a function that calls into other lower level functions) and each of
> those sub-operations might have a notion of status or error. If those
> sub-operations aren't represented with a child span then it is out of scope
> how their status is captured.

- now there might be multiple reasonable representations for a status and the
user becomes unclear which one to pick. In general I expect this to be mitigated
with documented conventions, API defaults, and the potentially canonicalizing
data anywhere in the processing pipeline. 

One place I neglected to go further was defining additional types of error data
format beyond string/int/Exception. This might encapsulate things such as
key/value pairs (ala structured logging) or more complex or niche status types
(COM IErrorInfo). There is nothing inherently problematic with them but I felt
the increased expressiveness was getting diminishing returns. One mitigation to
key/value pair data in particular would be to put auxilliary data in the span
attributes using some convention, for example "Error.UserName"="Bob" might be
used together with an error message string "Failed to find user {UserName}".
Another mitigation might be adding language specific overloads that handle
additional error types.

The current proposal only gives one string which can be used either for a 
freeform message or a textual status code. Adding a 2nd string to the

> StatusData would allow both to be collected SxS. This is another example of

Reviewer 2: let's spell this abbrev. out to be more inclusive for non-racing sport enthusiasts and non-native speakers, sparing them the Google lookup I just had to do 😉
```suggestion
StatusData would allow both to be collected side by side. This is another example of
```

Author: Interesting, I never knew "SxS" was related to racing! I probably first saw the acronym at work but its been too long now for me to trace the origin. Happy to change it : )

Reviewer 2: Well I didn't conduct any diligent research, the racing terms were just the first that popped up when I was googling for that term 😄

key value pairs would be overly onerous and error-prone for a common usage
scenario. If desired the SDK or exporter could destructure it, but that can be
determined independently from API design and I'd like to keep it out of scope.

**API using Span event semantic conventions** - Most of the rationale for attribute semantic
conventions also applies here, events are effectively another key-value store
mechanism. The timestamp that is attached to an event appears to hold little
value as status is probably produced at the same approximate time the span end
timestamp is recorded. Similar to attribute conventions it sounds like there is
precedent for storing some errors as events.

**Move the API to a non-core package** - It is possible to have the Tracing
API expose status using Attribute or Event APIs, and then have a 2nd library
that exposes a strongly-typed API that converts the status information into
attrbute/event updates. This adds some complexity over a strongly typed API
declared directly on Span but if we identify this as an area that needs to be
more decoupled/versionable than other Span tracing APIs perhaps it would be
valuable.

> **Represent an error message in addition to a string error name** - 

Reviewer 2: ```suggestion
**Represent an error message in addition to a string error name**
```
